### PR TITLE
Adding a script to fetch swagger spec from federation apiserver

### PR DESCRIPTION
**Pull request by nikhiljindal**
First step for https://github.com/kubernetes/kubernetes/issues/30541.
Next step is to generate docs like http://kubernetes.io/docs/api-reference/v1/definitions/ from this swagger spec.

cc @kubernetes/sig-cluster-federation @kubernetes/sig-api-machinery

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/30823)

### CODE REVIEW

File: hack/lib/util.sh [hunk 1]
```diff
@@ -325,6 +325,43 @@ kube::util::gv-to-swagger-name() {
...
 }
 
+
+# Fetches swagger spec from apiserver.
+# Assumed vars:
+# SWAGGER_API_PATH: Base path for swaggerapi on apiserver. Ex:
+# http://localhost:8080/swaggerapi.
+# SWAGGER_ROOT_DIR: Root dir where we want to to save the fetched spec.
+# VERSIONS: Array of group versions to include in swagger spec.
+kube::util::fetch-swagger-spec() {
```
Comment (nikhiljindal) [inline hack/lib/util.sh]: This function will become a lot simpler when we have swagger 2.0.
All logic to determine which paths to fetch will be gone. Just fetch /swagger.json
cc @mbohlool

**Comment by nikhiljindal** [comment]
cc @mbohlool

File: hack/update-federation-swagger-spec.sh [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    kube::etcd::cleanup
+
+    kube::log::status "Clean up complete"
+}
+
+trap cleanup EXIT SIGINT
+
+kube::golang::setup_env
+
+TMP_DIR=$(mktemp -d /tmp/update-swagger-spec.XXXX)
```
Comment (mbohlool) [inline hack/update-federation-swagger-spec.sh]: TMP_DIR is not used.

File: hack/update-federation-swagger-spec.sh [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+ETCD_HOST=${ETCD_HOST:-127.0.0.1}
+ETCD_PORT=${ETCD_PORT:-2379}
+API_PORT=${API_PORT:-8050}
+API_HOST=${API_HOST:-127.0.0.1}
```
Comment (mbohlool) [inline hack/update-federation-swagger-spec.sh]: API_HOST is not used? probably change --bind-address (and insecure-bind-address) from 127.0.0.1 to $API_HOST. Would you please fix update-swagger-spec.sh for both of these unused variables if applicable?

File: hack/update-federation-swagger-spec.sh [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+KUBELET_PORT=${KUBELET_PORT:-10250}
```
Comment (mbohlool) [inline hack/update-federation-swagger-spec.sh]: this look to be unused too, but I see common/shared scripts checking for it, so I am not that sure about this one.

File: hack/lib/util.sh (hunk 1 above)
Comment (mbohlool) [inline hack/lib/util.sh]: ack.

**Comment by mbohlool** [comment]
It would have been easier to review if the generated files were in a separate commit.

Only some (probably) unused variables in the script, otherwise LGTM.

File: hack/update-federation-swagger-spec.sh (hunk 2 above)
Comment (nikhiljindal) [inline hack/update-federation-swagger-spec.sh]: removed

File: hack/update-federation-swagger-spec.sh (hunk 3 above)
Comment (nikhiljindal) [inline hack/update-federation-swagger-spec.sh]: done. I see API_HOST being used that way in other scripts as well. Good catch :)

File: hack/update-federation-swagger-spec.sh (hunk 4 above)
Comment (nikhiljindal) [inline hack/update-federation-swagger-spec.sh]: yes its unused. removed

**Comment by nikhiljindal** [comment]
Sorry that was me being lazy :)
Will keep it in mind in future.

Have updated code as per comments. Have kept the recent changes as a separate commit :)
PTAL

**Comment by mbohlool** [comment]
np, Thanks.

**Comment by nikhiljindal** [comment]
Thanks, squashed commits

**Comment by nikhiljindal** [comment]
`test/integration/pods` failed which seems unrelated
@k8s-bot unit test this issue: #IGNORE

**Pull request merged**